modules/gapi/src/compiler/passes/transformations.cpp

```diff
@@ -0,0 +1,144 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2019 Intel Corporation
+
+#include "precomp.hpp"
+
+#include <ade/util/zip_range.hpp>
+#include <ade/graph.hpp>
+
+#include "api/gcomputation_priv.hpp"
+
+#include "compiler/gmodel.hpp"
+#include "compiler/gmodelbuilder.hpp"
+#include "compiler/passes/passes.hpp"
+#include "compiler/passes/pattern_matching.hpp"
+
+namespace cv { namespace gimpl { namespace passes {
+namespace
+{
+std::unique_ptr<ade::Graph> makeGraph(const cv::GComputation& c) {
+    // Generate ADE graph from expression-based computation
+    std::unique_ptr<ade::Graph> pG(new ade::Graph);
+    ade::Graph& g = *pG;
+
+    cv::gimpl::GModel::Graph gm(g);
+    cv::gimpl::GModel::init(gm);
+    cv::gimpl::GModelBuilder builder(g);
+    auto proto_slots = builder.put(c.priv().m_ins, c.priv().m_outs);
+
+    // Store Computation's protocol in metadata
+    cv::gimpl::Protocol p;
+    std::tie(p.inputs, p.outputs, p.in_nhs, p.out_nhs) = proto_slots;
+    gm.metadata().set(p);
+
+    return pG;
+}
+
+bool transform(ade::Graph& main, const std::unique_ptr<ade::Graph>& patternG,
+    const cv::GComputation& substitute) {
+    GModel::Graph gm(main);
+
+    // Note: if there are multiple matches, p must be applied several times (outside-scope work)
+    // 1. fina pattern in graph
+    auto match1 = findMatches(*patternG, gm);
+    if (!match1.ok()) {
+        return false;
+    }
+
+    // 2. build substitute graph inside the main graph
+    cv::gimpl::GModelBuilder builder(main);
+    const auto& proto_slots = builder.put(substitute.priv().m_ins, substitute.priv().m_outs);
+    Protocol p;
+    std::tie(p.inputs, p.outputs, p.in_nhs, p.out_nhs) = proto_slots;
+
+    // 3. match pattern ins/outs to substitute ins/outs
+    auto match2 = matchPatternToSubstitute(*patternG, main,
+        GModel::Graph(*patternG).metadata().get<Protocol>(), p);
+
+    // FIXME: from the state perspective it's better to validate matchings prior to applying the
+    //        transformations: graph can already be (partially) transformed
+    GAPI_Assert(match2.partialOk());
+
+    // 4. make substitution
+    performSubstitution(gm, match1, match2);
+    return true;
+}
+}  // anonymous namespace
+
+void checkTransformations(ade::passes::PassContext&,  // FIXME: context is unused here
+    const gapi::GKernelPackage& transformations,
+    std::vector<std::unique_ptr<ade::Graph>>& patterns)
+{
+    const auto& transforms = transformations.get_transformations();
+    const auto size = transforms.size();
+    if (0 == size) return;
```

Comment:
should be `0u`. as a side note, I'm not sure who thought that this is a good idea to put extra `u`, `UL`, etc. (and report compile warnings otherwise!) for cases when we compare a variable of a very specific known type and a literal... to my mind, literal should be automatically cast to the type of the variable (disregarding the mutual order of literal and variable)